### PR TITLE
Backport to 1.19.2 with Crafting Tweaks integration

### PR DESCRIPTION
**Pull request by DenisMasterHerobrine**
Hello! 👋 
Recently found this mod and came across the problem that there's no 1.19.2 version for this mod while it could fit very well into our new 1.19.2 modpack. 
So we've made a backport with Crafting Tweaks integration. We tried to be as less destructive as possible, making the 1.19.2 codebase to be similar with 1.20.1 ones.

However, if you find any concerns regarding code quality, please let me know, we'll try our best to solve the issues.

⚠️ **NOTE: I haven't found any options for GitHub allowing to automatically create a branch after merging PR, so the branch won't be overwritten. Current branch is set for the PR is set to `1.20`. Afaik after creating a new branch there will be an option to relocate the PR to it's own branch?** (it suggests me `1.20` branch, because I started backporting from this branch)

If it goes well, I have some plans to maintain this version for some time. If there will be some feature requests or some incompatibilities or issues with 1.19.2 version of course.

# PR Changelog:
- Backported to 1.19.2 
- Switched from YACL to Cloth Config API for Mod Menu integration (and the config handling)
- - YACL is unsupported for 1.19.2 at the moment of writing and somewhat missing docs (I realized that the docs for YACL2 was on GitHub, but not the actual mod wiki), it was reworked in 3.x versions, but the config handler still required a rewrite due to a major YACL3 changes in 1.20.1.
`If you want to support and use new features of YACL in your mod, you will have to drop support for 1.19.2. (from original changelog of YACL 2.2.0 for 1.19.2)`
- - Also Cloth Config API is shipped in most of modpacks and required by most popular Fabric mods, so in the most cases it shouldn't cause any performance decreasing by loading the embedded cloth-config library. If there's a newer version of Cloth Config, the newer ones will take priority and embedded ones will be ignored and Enhanced Workbenches will load as a standalone mod.
- - It's quite simpler. In my opinion.
- - However, language files are now a bit confusing, due to an attempt to support both 1.19.2 and 1.20.1 versions with different config handlers. Both `pt_br` and `en_us` jsons are adapted to changes making them independent from 1.19.2 to 1.20.1.
- Added Crafting Tweaks integration
- - Also adapted Project Table and Crafting Station rendering for the buttons. (Items rotating, spreading, clearing, etc.)
- - There's an issue with Balm's `ModelLayersAccessor` mixin throwing `AssertionError` on **each atlas and resource pack load** making it attempt to load resources **recursively** and crashing the game. Stopping the mixin from doing an `AssertionError` injection fixes these issues related to crashing on world load or loading into the game in dev environment and production environment (can be reproduced in a modpack with 433 mods) (similar and related issue to crash on world load in dev environment (related to https://github.com/TwelveIterationMods/CraftingTweaks/issues/193))
- Added a configuration rule in Gradle to stop duplicated Fabric API instances from being loaded in dev environment.

Should work with most mods if not all, let me know if any mod breaks this version of Enhanced Workbenches. 
Tested in an instance with 433 mods, works as expected.

# Screenshots
screenshot1
screenshot2
screenshot3

# tl;dr
Backport with no major gameplay changes, mainly reworking internal stuff and added Crafting Tweaks compatibility.
In case you wish to check out the PR without building from the sources, this [release can be found here](https://github.com/BehindTheScenery/EnhancedWorkbenches/releases/tag/1.1%2B1.19.2).

### CODE REVIEW

**Comment by Luligabi1** [comment]
Insane work, I cannot thank you two enough just for the initiative. I'm having some issues with my wi-fi atm, but I'll try my best to get this out within a week 🙂 

> Stopping the mixin from doing an AssertionError injection fixes these issues related to crashing on world load or loading into the game in dev environment and production environment

There might be a linguage barrier here, I don't get what you're trying to say here. How are you stopping the mixin?


> Afaik after creating a new branch there will be an option to relocate the PR to it's own branch? 

That's correct. I'll clone the 1.20 branch before merging and then change the target to there, nothing has to be done on y'all's side.

**Comment by DenisMasterHerobrine** [comment]
> There might be a linguage barrier here, I don't get what you're trying to say here. How are you stopping the mixin?

Sorry, my english skills are not perfect. I try my best to rephrase that. 
I have created my own local Balm fork to find out what's causing the issue and managed to find that removing the `ModelLayersAccessor` mixin fixes the crash described in the issue and I was able to load Crafting Tweaks in dev environment to test the PR code.

If I look into the mixin code, there's an `AssertionError` callback after accessing `ALL_MODELS`, seems to be something broken here, because once it's get's loaded, it throws assertion exception every time Minecraft tries to load resources and does a resource reloading cycle (the same as F3+T) infinitely and it stops once I close the game, if I remove the `throw new AssertionError();` line, then it works just fine and I am able to load the game and the world in dev environment.

Code reference:
https://github.com/TwelveIterationMods/Balm/blob/3a078a9baef5722c0eb443169526a303401cfa3f/shared/java/net/blay09/mods/balm/mixin/ModelLayersAccessor.java#L13-L16 (1.19.2 branch)

Not a mixin expert, but I hope this somehow helps investigating the issue with dev environments.

**Comment by Luligabi1** [comment]
> Sorry, my english skills are not perfect. 

No worries, it's not my native language either 🙂 

> If I look into the mixin code, there's an `AssertionError` callback after accessing `ALL_MODELS`

Afaik, you need this syntax for creating accessors of static fields, so it shouldn't be of any issue. Although I didn't find any place where the accessor is used, so it might be best their author remove it (but then again, it's a library, so other mods might use it)

I'm a bit skeptical of merging something on a version I don't plan to support for too long (hopefully in the next few months 1.20 modding will pick up steam) that I can't even test properly without jumping through all these hoops... it might be better to leave out CT support for now. Alternatively, I can upload a single-release for you to use on your modpack, although at the current state I wouldn't be able to give you too much support in case something goes wrong with it.

**Comment by DenisMasterHerobrine** [comment]
> I'm a bit skeptical of merging something on a version I don't plan to support for too long (hopefully in the next few months 1.20 modding will pick up steam) that I can't even test it properly... it might be better to leave out CT tweaks (or alternatively, have an unsupported, single-release with it).

As far as I tested it with removed mixin, it worked just fine. However, I understand that position about the CT. I can make it disabled by default in dev environment and once the issue will be resolved, re-enable it to test properly without any problems.

I understand that supporting on multiple versions is a bit tedious, but we're ready to maintain 1.19.2 branch atleast several months. (The modpack is heavily WIP and not ready for the production for now and we're in a way to port some other mods too)

Atleast I can maintain this project as a separated fork, since Apache License seems to be allowing to modify the code and use in modpacks if the major changes will be required to 1.19.2 branch with stated changes and included copyright and other legal stuff.

EDIT: However, I don't think that Enhanced Workbenches will recieve major updates and/or changes, probably single-release can be a solution for now, just simply to use in a modpack or other modpacks if players wish to.

**Comment by Luligabi1** [comment]
> once the issue will be resolved

I doubt this issue will be solved anytime soon, if ever on 1.19.2... the [author has said before this is a longstanding issue that they do not know how to fix nor how to trigger](https://github.com/TwelveIterationMods/CraftingTweaks/issues/193#issuecomment-1628545699).

> However, I don't think that Enhanced Workbenches will recieve major updates and/or changes, probably single-release can be a solution for now, just simply to use in a modpack or other modpacks if players wish to.

Yeah, I don't plan any new content for the foreseeable future, much less for 1.19.2. If you both are fine with it, I can do the following:

- Merge this on branch `1.19.2-ct`
- Do the required (if any) changes
- Release that as `1.0-BETA+1.19.2` with clear indications it's an unsupported build
- Create a new `1.19.2` branch, same as `1.19.2-ct`, but without CT support
- Release that as `1.0+1.19.2`, with regular support from me

While the license allows forks, I generally find them somewhat of a pain as often users will come to me for support of some shortlived fork... if we can come to an agreement, it'll be much better for both of us 😉

**Comment by DenisMasterHerobrine** [comment]
> If you both are fine with it.

Atleast I'm fine with it, VeiTr is probably will be fine too (different time zones).

> While the license allows forks, I generally find them somewhat of a pain as often users will come to me for support of some shortlived fork... if we can come to an agreement, it'll be much better for both of us 😉

I think it would great for both of us too making two separated versions.
If any changes our modpack will require (some weird mod incompatibilities if we find any later), probably I will do another PR to `1.19.2-ct` branch, while this version is going be unsupported by you.

**Comment by Luligabi1** [comment]
> > If you both are fine with it.
> 
> Atleast I'm fine with it, VeiTr is probably will be fine too (different time zones).
> 
> > While the license allows forks, I generally find them somewhat of a pain as often users will come to me for support of some shortlived fork... if we can come to an agreement, it'll be much better for both of us 😉
> 
> I think it would great for both of us too making two separated versions. If any changes our modpack will require (some weird mod incompatibilities if we find any later), probably I will do another PR to `1.19.2-ct` branch, while this version is going be unsupported by you.

Fine by me 🙂 I'll tag you once it's released

**Comment by Luligabi1** [comment]
@DenisMasterHerobrine it's here: https://legacy.curseforge.com/minecraft/mc-mods/enhanced-workbenches/files/4682408

Thanks again for the PR, feel free to open others and contact me via Discord 🙂

**Comment by DenisMasterHerobrine** [comment]
Thank you for uploading this. 👍 
Balm `4.5.7` and Crafting Tweaks `15.1.8` works as intended with Enhanced Workbenches `1.0-BETA+1.19.2` in the modpack without any crashes. If something arises, I'll let you know.
